payload_builder refactor

tydangelo18 opened the pull request:
payload builder refactor. No more chaining. Now just pass in required params for tx or msg and call build function.

KwilLuke commented on .vscode/settings.json:
```diff
@@ -0,0 +1,6 @@
+{
```
please git ignore your vscode settings

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -143,11 +143,14 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
-      this,
-      PayloadType.DEPLOY_DATABASE
-    ).chainId(this.chainId);
+  public dbBuilder(): DB<T> {
```
You can remove this, bcz dbBuilder is deprecated

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -157,10 +160,13 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
-    return DBBuilderImpl.of<PayloadType.DROP_DATABASE, T>(this, PayloadType.DROP_DATABASE).chainId(
-      this.chainId
-    );
+  public dropDbBuilder(): DB<T> {
```
You can also remove this, as it is deprecated (and we are now beyond the removal period)

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -231,16 +237,16 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+   
+    let tx = DB.create(this, PayloadType.DEPLOY_DATABASE, {
+      description: deployBody.description || '',
+      payload: () => deployBody.schema,
```
why are you making the schema into a callback function? Why not just pass `deployBody.schema`?

KwilLuke replied:
Just reminder: we can get rid of the CB

KwilLuke commented on src/db/db.ts:
```diff
@@ -0,0 +1,486 @@
+  private payload?: () => CompiledKuneiform | DropDbPayload;
+  private identifier?: Uint8Array;
+  private signatureType?: AnySignatureType;
+  private chainId?: string;
```
What would be the scenario that any of these arguments would not be required? I think for any deploy/drop, each of these args would be required.

The only two args that are optional are the description (if the user does not pass anything to `actionBody.description` and the `nonce`, in which case, you have the get the nonce from the network.)

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -231,16 +237,16 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+      nonce: deployBody.nonce && deployBody.nonce,
+    });
 
     const transaction = await tx.buildTx();
```
Why the separate build method?

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -263,16 +269,16 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
-    }
+    let tx = DB.create(this, PayloadType.DROP_DATABASE, {
+      description: dropBody.description || '',
+      payload: () => ({ dbid: dropBody.dbid }),
```
Same comment on `deploy` - why are you making this a cb function?

KwilLuke replied:
Reminder about CB here.

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -231,16 +237,16 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+      signer: kwilSigner.signer,
+      signatureType: kwilSigner.signatureType,
+      chainId: this.chainId,
+      nonce: deployBody.nonce && deployBody.nonce,
```
Couldn't you just do

```ts
nonce: deployBody.nonce
```

if `deployBody.nonce` is defined, it will be assigned to `nonce`. If it is undefined, it will be passed to create as `undefined` (for which you should have logic to handle in `DB` because `nonce` is an optional param)

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -263,16 +269,16 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+      signer: kwilSigner.signer,
+      signatureType: kwilSigner.signatureType,
+      chainId: this.chainId,
+      nonce: dropBody.nonce && dropBody.nonce,
```
Same nonce comment here as on line 248 (the deploy method)

KwilLuke commented on src/utils/validation.ts:
```diff
@@ -0,0 +1,22 @@
+ * @param values An object containing field names and their corresponding values.
+ * @param errorMessageTemplate A function to generate error messages dynamically.
+ */
+export const validateFields = <T extends Record<string, unknown>>(
```
Can you make this function way simpler.

KwilLuke replied:
Also, why the generic if it will just be `Record<string, unknown>`? Couldn't you just assign that type to `values`?

KwilLuke commented on src/payload/payload.ts:
```diff
@@ -0,0 +1,283 @@
+    const resolvedPayload = await this.resolvePayload();
+
+    // ensure required fields are not null or undefined
+    const { signer, payloadType, identifier, signatureType, chainId } = validateFields(
```
I like what you did here with `validateFields`. Could you add some unit tests for this function?

tydangelo18 replied:
sure. We should talk through this Monday. There are some things I'm not understanding

KwilLuke commented on src/utils/validation.ts:
```diff
@@ -0,0 +1,22 @@
+      ) as NonNil<T[typeof fieldName]>;
+    }
+    return validatedFields;
+  };
```
I think all of the type assertions is what makes this hard to read. If you could find a way to do this without the type assertions, that would be solid.

tydangelo18 replied:
this is an interesting problem. Still working through this

KwilLuke reviewed (commented):
Partial review... Didn't get through the whole thing, but just some food for thought.

KwilLuke commented on testing-functions/gavin.ts:
```diff
@@ -0,0 +1 @@
+console.log("Gavin=========");
```
want to remove this before merge?

KwilLuke commented on package.json:
```diff
@@ -15,6 +15,7 @@
     "LICENSE"
   ],
   "scripts": {
+    "start": "ts-node testing-functions/scratchpad.js",
```
Kind of a weird script for "start" (from the prespective of someone that views package.json for the first time).

Change this script to "scratchpad"?

KwilLuke commented on src/utils/typeHelpers.ts:
```diff
@@ -0,0 +1,16 @@
+/**
```
Not sure why we originally had this, but if you don't call anywhere, you can delete.

KwilLuke commented on src/utils/objects.ts:
```diff
@@ -1,72 +1,121 @@
+      ) as NonNil<T[typeof fieldName]>;
+    }
+    return validatedFields;
+  },
```
I am trying to cut down on all the type casting (hard to read and potentially error prone).

How would does this approach look to you?


```suggestion

  validateFields: <T extends Record<string, any>>(  
    values: T,  
    errorMessageTemplate: (fieldName: keyof T) => string  
  ): { [K in keyof T]: NonNil<T[K]> } => {  
    for (const key in values) {  
      objects.requireNonNil(  
        values[key],  
        errorMessageTemplate(key)  
      );  
    } 
    return values;  
  },
```

KwilLuke commented on src/builders/action_builder.ts:
```diff
@@ -29,6 +29,7 @@ const TXN_BUILD_IN_PROGRESS: ActionInput[] = [];
 /**
  * `ActionBuilderImpl` class is an implementation of the `ActionBuilder` interface.
  * It helps in building and transactions to execute database actions on the Kwil network.
+ * @deprecated Use `Action` class instead`.
```
Can we just remove this file?

tydangelo18 replied:
will remove once we're finalized in review

KwilLuke commented on src/builders/db_builder.ts:
```diff
@@ -20,6 +20,7 @@ import { AllPayloads, CompiledKuneiform, DbPayloadType, DropDbPayload } from '..
 /**
  * `DBBuilderImpl` class is an implementation of the `DBBuilder` interface.
  * It creates a transaction to deploy a new database on the Kwil network.
+ * @deprecated Use `DB` class instead`.
```
Can we just remove this file?

tydangelo18 replied:
will remove once we're finalized in review

KwilLuke commented on src/builders/payload_builder.ts:
```diff
@@ -24,6 +24,7 @@ import { AllPayloads, UnencodedActionPayload } from '../core/payload';
 /**
  * PayloadBuilderImpl is the default implementation of PayloadBuilder. It allows for building transaction and call payloads that can be sent over GRPC.
  * See the proto files for more information on the structure of the payloads. {@link https://github.com/kwilteam/proto/tree/main/kwil/tx/v1}
+ * @deprecated Use `Payload` class instead`.
```
Can we just remove this file?

tydangelo18 replied:
will remove once finalized in review

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+    this.description = objects.requireNonNil(
+      options.description,
+      'Description cannot be null or undefined.'
+    );
```
Would this be a place you could use `validateRequiredFields`? https://github.com/kwilteam/kwil-js/blob/52b2e5a558b26a2a5ba5730b4fe31ab30fe098a7/src/utils/objects.ts#L114

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+    const { dbid, actionName, preparedActions, modifiers } = await this.checkSchema(
+      cached as ActionInput[]
+    );
+
```
Again, it would be good here if you could find a way to avoid type casting.

In this case, the fix is simple:

On 132, change:

```diff
-    const { dbid, actionName, preparedActions, modifiers } = await this.checkSchema(
-      cached as ActionInput[]
-    );
+    const { dbid, actionName, preparedActions, modifiers } = await this.checkSchema(cached);
```

On the `checkSchema` definition, change:

```diff
 private async checkSchema(actions: ActionInput[]): Promise<CheckSchema> {
+  private async checkSchema(actions?: ActionInput[]): Promise<CheckSchema> {
     // throw runtime errors is dbid or action name is null or undefined
     const dbid = objects.requireNonNil(
```

`checkSchema` can already account for `actions` being optional, because of this section here: https://github.com/kwilteam/kwil-js/blob/52b2e5a558b26a2a5ba5730b4fe31ab30fe098a7/src/action/action.ts#L329-L338

tydangelo18 replied:
lgtm

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+      this.actionName,
+      'name is required to execute or call an action or procedure.'
+    );
+
```
What do you think about moving this `objects.requiredNonNil` checks to the constructor? For other properties on this class, the check is done in the constructor... would it make the most sense to move this there too?

tydangelo18 replied:
I agree. This is something I thought about. Almost did it and figured buildTx needed to additionally check for it

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+          public: procedureSchema?.public as boolean,
+          parameters: (procedureSchema?.parameters?.map((p) => p.name) as string[]) || [],
+          modifiers: procedureSchema?.modifiers as string[],
+        };
```
Let's try to remove the typecasting here... Here is how I think you can do it:

First, chage `validateRequiredFields` to return. In `utils/objects`:

```diff
    * @throws Error if any field is explicitly provided but null.
    */
-  validateRequiredFields<T>(options: T, fields: (keyof T)[]): void {
+  validateRequiredFields<T>(options: T, fields: (keyof T)[]): T {
     fields.forEach((field) => {
       if (options[field] === undefined && options[field] === null) {
         throw new Error(`${String(field)} must not be null.`);
       }
     });
+    return options;
   },
-};
+};
```

Then change `checkSchema` in `src/action/action.ts`:

```diff
+    const validated = objects.validateRequiredFields(
+      foundActionOrProcedure,
+      ['name', 'public']
+    );
+
     const execSchema: ExecSchema = actionSchema
       ? {
-          name: actionSchema.name,
-          public: actionSchema.public,
+          name: validated.name,
+          public: validated.public,
           parameters: actionSchema.parameters,
           modifiers: actionSchema.modifiers,
         }
       : {
           // if we have reached this point and actionSchema is null, then we know that procedureSchema is not null.
-          name: procedureSchema?.name as string,
-          public: procedureSchema?.public as boolean,
-          parameters: (procedureSchema?.parameters?.map((p) => p.name) as string[]) || [],
-          modifiers: procedureSchema?.modifiers as string[],
+          name: validated.name,
+          public: validated.public,
+          parameters: (procedureSchema?.parameters?.map((p) => p.name)) || [],
+          modifiers: procedureSchema?.modifiers,
         };
```

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+  private chainId: string;
+  private description: string;
+  private signer?: SignerSupplier;
+  private identifier?: HexString | Uint8Array;
```
I think identifier should be just `uint8Array`. (Previously, the SDK managed all identifiers internally as Uint8array). When would it be HexString?

KwilLuke replied:
(Btw, when using the `kwilSigner` class, identifier will always be `Uint8Array`)

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+      challenge: this.challenge,
+      signature: this.signature,
+      chainId: chainId,
+      identifier: identifier as Uint8Array, // Fix this. Seems hacky...
```
Yep, remove type cast.

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+      // if there are nilArgs, then the first element in the array is the nilArgs.
+    };
+
+    let msg = await Payload.createTx(this.kwil, {
```
Don't need to `await`.

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+
+    // retrieve the schema and run validations
+    const { dbid, actionName, preparedActions, modifiers } = await this.checkSchema(
+      cached as ActionInput[]
```
No type casting!

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+    const { signer, chainId, identifier, signatureType } = objects.validateFields(
+      {
+        signer: this.signer,
+        chainId: this.chainId,
```
chainId is needed in both `buildMsg` and `buildTx`. Maybe move the `chainId` check to the constructor?

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+    }
+
+    // resolve if a signer is or is not specified. Only actions with `mustsign` require a signer.
+    const signer = this.signer ? this.signer : null;
```
`mustsign` is an old thing. We can prob just remove this and pass `this.signer` to `Payload.createTx`.

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+        ...msg,
+        signer: signer,
+        signatureType: signatureType,
+        identifier: identifier as unknown as Uint8Array,
```
Booooo. No type casting!!! 😭

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+
+    // if a signer is specified, add the signer, signature type, identifier, and description to the message
+    if (signer) {
+      const identifier = await Promisy.resolveOrReject(this.identifier as Nillable<string>);
```
identifier should be `Uint8Array`, not string.

Also, 
![9bqtu0](https://github.com/user-attachments/assets/e5629584-f205-4cfc-82d9-38320201accc)

KwilLuke replied:
Also, why do you need to `Promisy.resolveOrReject` on `this.identifier`? `this.identifier` is not a promise?

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+    // if a signer is specified, add the signer, signature type, identifier, and description to the message
+    if (signer) {
+      const identifier = await Promisy.resolveOrReject(this.identifier as Nillable<string>);
+      const signatureType = await Promisy.resolveOrReject(this.signatureType);
```
Same here - no need to `Promisy.resolveOrReject`. This.signatureType does not need to be resolved.

I can explain `Promisy` was originally for offline, but the main point is: Bcz most args to it are now internal (and the args you are passing are resolved), they aren't needed anymore.

KwilLuke commented on src/payload/payload.ts:
```diff
@@ -0,0 +1,306 @@
+import { strings } from '../utils/strings';
+
+export interface PayloadOptions {
+  payload: (() => AllPayloads) | AllPayloads;
```
I actually don't think we accept a CB in any of our public APIs anymore, so you can prob change this to just:

```
payload: AllPayloads
```

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -193,21 +153,28 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
-
-    if (actionBody.nonce) {
-      tx = tx.nonce(actionBody.nonce);
+      const inputs = (actionBody.inputs as ActionInput[]).every(
```
Type casting: ![9bqtu0](https://github.com/user-attachments/assets/e5629584-f205-4cfc-82d9-38320201accc)

KwilLuke replied:
Seriously though, try to write without `as`.

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -193,21 +153,28 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+      )
+        ? (actionBody.inputs as ActionInput[])
+        : new ActionInput().putFromObjects(actionBody.inputs as Entries[]);
+      tx = Object.assign(tx, {
```
IMO, it would be better to do this validation before calling `createTx` and then just include the inputs in `createTx`. That way, you avoid the awkward `Objects.assign`.

KwilLuke commented on src/action/action.ts:
```diff
@@ -0,0 +1,440 @@
+       *  extend the `msg` object by adding the new properties while preserving the existing type
+       */
+      //
+      msg = Object.assign(msg, {
```
Object.assign is quite awkward (also, potentially error prone if we add validation logic in the constructor). Can you find an alternative way?

KwilLuke reviewed (changes requested):
## Initial Review Pt. 1

My review checklist:

- [x] src
  - [x] action
    - [x] action.ts
  - [x] api_client
  - [x] builders
    - [x] action_builder.ts
    - [x] db_builder.ts
    - [x] payload_builder.ts
  - [x] client
    - [x] node
      - [x] nodeKwil.ts
    - [x] web
      - [x] webKwil.ts
    - [x] kwil.ts
  - [x] core
    - [x] action.ts
    - [x] database.ts
  - [x] db
    - [x] db.ts
  - [x] funder
    - [x] funder.ts
  - [x] payload
    - [x] payload.ts
  - [x] utils
    - [x] objects.ts
    - [x] typeHelpers.ts
    - [x] types.ts
  - [x] testing-functions
    - [x] gavin.ts
    - [x] scratchpad.js
    - [x] test.js
  - [x] tests
    - [x] validation
      - [x] kwil.test.ts
      - [x] validation.test.js
  - [x] package-lock.json
  - [x] package.json

I am about 60%ish complete. Posting now so you can start seeing some of the feedback.

Overall, I like the new architecture and where this is going. I still think we have room to simplify and clean up.

Also, PLEASE try to avoid type assertion. I've tried to call it out in a number of places in the review, however, please do your own review and try to clean up where possible.

I'll post a part 2 later.

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -193,22 +153,26 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+      )
+        ? (actionBody.inputs as ActionInput[])
+        : new ActionInput().putFromObjects(actionBody.inputs as Entries[]);
     }
```
Type casting. You can change to this:


```suggestion
      inputs = actionBody.inputs.every(
        (item) => item instanceof ActionInput
      )
        ? (actionBody.inputs)
        : new ActionInput().putFromObjects(actionBody.inputs);
```

KwilLuke replied:
The type casting is still not removed here.

KwilLuke commented on src/client/web/webKwil.ts:
```diff
@@ -105,29 +106,31 @@ export class WebKwil extends Kwil<EnvironmentType.BROWSER> {
           ? (actionBody.inputs as ActionInput[])
           : new ActionInput().putFromObjects(actionBody.inputs as Entries[]);
-      msgBuilder = msgBuilder.concat(inputs);
     }
```
Type assertion here.

tydangelo18 replied:
@KwilLuke need some guidance with this type assertion. I've tried a lot of troubleshooting but can't seem to make it work by removing as Entries[]. Which is odd, because if you look in kwil.execute(), you'll see that doesn't need type assertions for the inputs logic...

KwilLuke commented on src/db/db.ts:
```diff
@@ -0,0 +1,512 @@
+
+export interface DBOptions {
+  signer: SignerSupplier;
+  payload: () => CompiledKuneiform | DropDbPayload;
```
No cb

KwilLuke commented on src/db/db.ts:
```diff
@@ -0,0 +1,512 @@
+  public readonly kwil: Kwil<T>;
+  public payloadType: DeployOrDrop;
+  public signer: SignerSupplier;
+  public payload: () => CompiledKuneiform | DropDbPayload;
```
Also flagging CB here for you

KwilLuke commented on src/db/db.ts:
```diff
@@ -0,0 +1,512 @@
+      options.signatureType,
+      'Signature type cannot be null or undefined. Please specify signature type.'
+    );
+    this.chainId = objects.requireNonNil(options.chainId, 'Chain ID cannot be null or undefined.');
```
Use `validateRequiredFields`?

KwilLuke commented on src/db/db.ts:
```diff
@@ -0,0 +1,512 @@
+      objects.requireNonNil(
+        payloadType,
+        'payloadType is required for DbBuilder. Please pass a valid PayloadType. This is an internal error, please create an issue.'
+      );
```
Move this to constructor?

KwilLuke commented on src/db/db.ts:
```diff
@@ -0,0 +1,512 @@
+    this.payload =
+      typeof ensuredPayload !== 'function'
+        ? () => ensuredPayload
+        : (ensuredPayload as () => CompiledKuneiform | DropDbPayload);
```
No need to do lazy eval anymore. Just an object.

KwilLuke commented on src/payload/payload.ts:
```diff
@@ -0,0 +1,306 @@
+  description?: string;
+  nonce?: number;
+  challenge?: string;
+  signature?: Signature<BytesEncodingStatus.BASE64_ENCODED>;
```
All of these being optional looks incorrect, because some instances (anything that requires a transaction), many of these are actually not optional.

This makes me think we need to separate building the call payload from building the transaction payload (i.e., separate classes entirely). WDYT

KwilLuke commented on src/payload/payload.ts:
```diff
@@ -0,0 +1,306 @@
+   *
+   * @returns {AllPayloads} - A promise that resolves to the provided payload object.
+   */
+  private async resolvePayload(): Promise<AllPayloads> {
```
Dont need `resolvePayloads` anymore.

KwilLuke commented on tests/unitTests/validation/validation.test.js:
```diff
@@ -0,0 +1,112 @@
+    });
+
+    // PROVIDE WALLET/SIGNER BUT NO ADDRESS/IDENTIFIER
+    const kSigner = new kwiljs.KwilSigner(wallet);
```
This should fail when you create the `KwilSigner`. Does it not?

tydangelo18 replied:
it fails, but I assert/expect the failure in the test.

KwilLuke reviewed (changes requested):
Few more comments. Looking good.

I think the main thing we should try to improve is to breakdown `payload.ts` a bit more. I am concerned that too many of the parameters are optional, when for many use cases they should not be optional. This makes me think we need something separate for transactions and call requests (msgs).

KwilLuke commented on src/client/web/webKwil.ts:
```diff
@@ -28,195 +20,14 @@ export class WebKwil extends Kwil<EnvironmentType.BROWSER> {
    */
   public async call(
-    actionBody: ActionBody,
+    actionBody: Message | ActionBody,
```
We can remove `Message` here. Only `ActionBody` is needed.

KwilLuke commented on src/client/node/nodeKwil.ts:
```diff
@@ -28,226 +20,47 @@ export class NodeKwil extends Kwil<EnvironmentType.NODE> {
    */
   public async call(
-    actionBody: ActionBodyNode,
+    actionBody: Message | ActionBodyNode,
```
Can also remove `Message` here.

KwilLuke commented on src/client/node/nodeKwil.ts:
```diff
@@ -28,226 +20,47 @@ export class NodeKwil extends Kwil<EnvironmentType.NODE> {
+        // reset the temporary cookie
+        if (tempCookie) {
+          this.resetTempCookie(tempCookie);
+        }
```
Does this work? You're just setting the cookie twice?

KwilLuke replied:
Actually, i don't think this will work. Any cookie the user passes will just be overridden.

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -357,4 +324,203 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+   */
+  protected async baseCall(
+    actionBody: Message | ActionBodyNode,
+    kwilSigner?: KwilSigner,
```
This needs to be `ActionBody`, not `ActionBodyNode`

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -181,34 +150,38 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
    */
   public async execute(
-    actionBody: ActionBody,
+    actionBody: ActionBodyNode,
```
No, just use `ActionBody` here. This needs to be for both web and nodejs

KwilLuke commented on src/core/database.ts:
```diff
@@ -29,9 +34,10 @@ export interface DropBody {
+  tables: ReadonlyArray<CompiledTable>;
+  procedures: ReadonlyArray<CompiledProcedure>;
+  foreign_calls: ReadonlyArray<CompiledForeignProcedure>;
+  type?: 'encodeable_database';
```
Why? This feels quite weird to put here because `type` is not a database property that kwild will accept.

KwilLuke commented on src/core/payload.ts:
```diff
@@ -1,56 +1,73 @@
 export interface DropDbPayload {
-    dbid: string
+  dbid: string;
+  type?: 'drop_db_payload';
```
Again: why?

KwilLuke commented on src/core/payload.ts:
```diff
@@ -1,56 +1,73 @@
+  extensions: PartialNillable<Extension>[] | null;
+  procedures: PartialNillable<CompiledProcedure>[] | null;
+  foreign_calls: PartialNillable<CompiledForeignProcedure>[] | null;
+  type?: 'compiled_kuneiform';
```
Feels weird here because `type` isn't a valid db type in kwild.

KwilLuke commented on src/core/payload.ts:
```diff
@@ -59,48 +76,49 @@ export interface CompiledKuneiform {
-    amount: string;
+  to: T extends BytesEncodingStatus.BASE64_ENCODED ? string : Uint8Array;
+  amount: string;
+  type?: 'transfer_payload';
```
Again, we don't need this! We have `PayloadType` for a reason.

KwilLuke commented on src/db/db.ts:
```diff
@@ -0,0 +1,491 @@
+    const signatureType = await Promisy.resolveOrReject(
+      this.signatureType,
+      'signature type cannot be null or undefined. please specify a signature type.'
+    );
```
You don't need Promisy.

KwilLuke commented on src/payload/payloadMsg.ts:
```diff
@@ -0,0 +1,127 @@
+ * `PayloadMsg` class creates a call message payload that can be sent over GRPC.
+ */
+export class PayloadMsg {
+  public payload: AllPayloads;
```
Does this actually need to be `AllPayloads`? I don't think so...

You're only going to use this for actions.

KwilLuke commented on src/payload/payloadMsg.ts:
```diff
@@ -0,0 +1,127 @@
+  ): Promise<Message> {
+    // rlp encode the payload
+    const encodedPayload = kwilEncode(
+      msg.body.payload as UnencodedActionPayload<PayloadType.CALL_ACTION>
```
Please, let's try to avoid type assertion at all costs.

KwilLuke reviewed (changes requested):
Some comments, mostly all minor. The biggest issue in this PR is how you were handling the NodeJS cookie.

I implemented some (but not all) of the changes needed as a did the review. Will open a PR and leave you to implement the remaining changes.

tydangelo18 commented:
@KwilLuke a couple of articles supporting why I chose the alternative I did for getting rid of the type assertions in kwil.ts

1) https://www.allthingstypescript.dev/p/avoid-using-type-assertions-in-typescript

2) https://blog.logrocket.com/how-to-use-type-guards-typescript/

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -357,4 +329,224 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
+    const name = !actionBody.name && actionBody.action ? actionBody.action : actionBody.name;
+
+    let inputs: ActionInput[] | undefined;
+    if (actionBody.inputs) {
+      inputs = this.isFirstElementActionInput(actionBody.inputs)
+        ? actionBody.inputs
+        : new ActionInput().putFromObjects(actionBody.inputs);
+    }
+
```
Nice approach!

KwilLuke commented:
> @KwilLuke a couple of articles supporting why I chose the alternative I did for getting rid of the type assertions in kwil.ts
> 
> 1. https://www.allthingstypescript.dev/p/avoid-using-type-assertions-in-typescript
> 2. https://blog.logrocket.com/how-to-use-type-guards-typescript/

Nice!

KwilLuke reviewed (approved):
lgtm to me. going to push some minor cleanups.

Please squash all the commits down into one msg (formatted from `npm run commit`).

KwilLuke committed:
refactor: remove builders, improve separation

Remove builder classes for managing procedures/actions/deploy/drop. Improve logical separation
between transactions and messages.

Co-Authored-By: Ty D'Angelo <tydangelo18@gmail.com>
